### PR TITLE
add risks handler

### PR DESCRIPTION
**Pull request by ghost**
Patch for ticket #29 

I have not tested it tho. 

But I'm pretty sure it works.
If not, then 1 or 2 adjustments max should do the trick.

Thanks !

### CODE REVIEW

File: covbot.py [hunk 1]
```diff
@@ -597,6 +599,25 @@ def _short_location(self, location: str, length=int(12)) -> str:
         c.format = "org.matrix.custom.html"
         await e.respond(c, markdown=True, allow_html=True)
 
+    @command.new('risks', help=HELP['risks'][1])
+    @command.argument("risks", pass_raw=True, required=True)
```
Comment (bertiebaggio) [inline covbot.py]: -    @command.argument("risks", pass_raw=True, required=True)
    +    @command.argument("age", pass_raw=True, required=True)

File: covbot.py [hunk 2]
```diff
@@ -597,6 +599,25 @@ def _short_location(self, location: str, length=int(12)) -> str:
...
+    async def risks_handler(self, event: MessageEvent, age: int) -> None:
+        if age < 0 or age > 110:
```
Comment (bertiebaggio) [inline covbot.py:595]: Hinting isn't enough, you need to explicitly `int(age)` if you want that.

There doesn't seem to be a way to specify an explicit type in the decorator, unless I missed it:
https://github.com/maubot/maubot/blob/master/maubot/handlers/command.py#L319

File: covbot.py [hunk 3]
```diff
@@ -597,6 +599,25 @@ def _short_location(self, location: str, length=int(12)) -> str:
...
+            await self._respond(event, "Age must be between 0 and 110")
+            return
+        death_rate = -0.0012822 + (0.00000334593 * age ** 2) + ((2.7762 * 10 ** -15) * age ** 7)
+        intensive_care_rate = (1 / math.pi) + -0.358209 / (1 + 0.0827213 * math.exp(0.035025 * age))
+        hospitalization_rate = -0.0353956 - age * -0.00576217
+        survival_rate = 1 - death_rate
+        pf = lambda x: round(x * 100)
+        response = """
```
Comment (bertiebaggio) [inline covbot.py]: This gives an interestingly-indented message:

covbotrisks


I would suggest not indenting for consistency in output with `!cases`, but that's @pwr22's call overall.

**Review by bertiebaggio — CHANGES_REQUESTED** [review]
Update the argument (risks → age), use as int, check with @pwr22 about the indentation/formatting of the message.

**Comment by bertiebaggio** [comment]
This is an interesting feature! I've tested it on my own bot with enough changes to get it to work.

Just to be clear, I have no particular authority here- I just reviewed the bits that jumped out at me as I had the time while I was working on my own PR. So don't take anything I say as gospel, @dctremblay ! @pwr22, I hope you don't mind my jumping in here uninvited :)

**Comment by pwr22** [comment]
I'm happy for any help 🙂.

File: covbot.py (hunk 1 above)
Comment (ghost) [reply covbot.py]: Uh oh make sense. I fix this right now, thanks

File: covbot.py (hunk 2 above)
Comment (ghost) [reply covbot.py:595]: Thank you very much ! I was wondering

File: covbot.py [hunk 4]
```diff
@@ -609,7 +609,7 @@ def _short_location(self, location: str, length=int(12)) -> str:
         intensive_care_rate = (1 / math.pi) + -0.358209 / (1 + 0.0827213 * math.exp(0.035025 * age))
         hospitalization_rate = -0.0353956 - age * -0.00576217
         survival_rate = 1 - death_rate
-        pf = lambda x: round(x * 100)
+        pf = lambda x: round(x * 100, 1)
```
Comment (bertiebaggio) [inline covbot.py]: You can format directly as percentage and omit the lambda, eg `{:.1%}` will render 0.3333333333 as `33.3%` :)

**Review by ghost — COMMENTED** [review]
Thank you very much for the review. I was too lazy to build development environment for a few line patch aha. I'm really sorry for the inconvenience tho. I'm fixing points right now.

**Comment by bertiebaggio** [comment]
My pleasure. I have a very simple and quick build+upload setup with `entr` so it's very easy for me to test this- no inconvenience at all!

----

In case anyone's wondering:

```
$ ls ~/code/covbot/covbot.py ~/code/covbot/maubot.yaml | entr ./covbot_deploy.sh

$ cat covbot_deploy.sh

python3 -m maubot.cli build /home/robert/code/covbot
python3 -m maubot.cli upload dev.shortestpath.covbot-v0.0.13rc0.mbp
```

the `ls` line watches for changes to `covbot.py` or `maubot.yaml`. I could make it smarter and one-lined and pass the built .mbp automatically, but that's not really needed most of the time!

File: covbot.py (hunk 4 above)
Comment (ghost) [reply covbot.py]: ohh nice didnt knew this trick ! i am doing this now

**Comment by ghost** [comment]
:D thank you very much

**Comment by ghost** [comment]
ready to rereview

File: covbot.py (hunk 3 above)
Comment (ghost) [reply covbot.py]: yup it had inconsistent formatting. fixed that !

File: covbot.py [hunk 5]
```diff
@@ -597,6 +599,24 @@ def _short_location(self, location: str, length=int(12)) -> str:
...
+        # source : https://www.desmos.com/calculator/v0zif7tflm
+        age = int(age)
+        if age < 0 or age > 110:
+            await self._respond(event, "Age must be between 0 and 110")
+            return
+        death_rate = max(0, -0.00186807 + 0.00000351867 * age ** 2 + (2.7595 * 10 ** -15) * age ** 7)
+        intensive_care_rate = max(0, -0.0572602 - -0.0027617 * age)
+        hospitalization_rate = max(0, -0.0730827 - age * -0.00628289)
+        survival_rate = 1 - death_rate
+        response = '\n'.join([
```
Comment (bertiebaggio) [inline covbot.py]: I would omit the manual newline, and have changed the wording very slightly:


```suggestion
        response = (
            "Based on reports, infection with SARS-CoV-2 in a {} year old has an average {:.1%} survival chance;"
            " with an average risk of hospitalization of {:.1%}, risk of needing intensive care of {:.1%},"
            " and {:.1%} chance of death."
        ).format(age, survival_rate, hospitalization_rate, intensive_care_rate, death_rate)
```

Which looks like:

covrisks2

**Comment by bertiebaggio** [comment]
I have made a further minor formatting suggestion

(I haven't reviewed the calculations)

File: covbot.py (hunk 5 above)
Comment (ghost) [reply covbot.py]: * risk of death lol

Comment (ghost) [reply covbot.py]: noice. its very clean code. thanks !

**Comment by pwr22** [comment]
I'll have to take a look at `entr` some time :thinking: .

**Review by pwr22 — APPROVED** [review]
LGTM :+1:

**Comment by ghost** [comment]
nice thanks :D

**Comment by pwr22** [comment]
Np, thank you for adding it!

**Comment by bertiebaggio** [comment]
> I'll have to take a look at `entr` some time  .

I must document the things I use it for :smiley:  If you are making changes to files and want to do something or see something resulting from that but don't want/need to set up CI/CD it's gold!

**Pull request merged**